src/librustc/session/filesearch.rs

```diff
@@ -61,17 +61,29 @@ impl<'a> FileSearch<'a> {
         make_target_lib_path(self.sysroot, self.triple)
     }
 
-    pub fn search<F>(&self, mut pick: F)
+    // If this function is called repeatedly on one or more directories,
+    // `cached_file_paths` avoids the need to re-traverse that directory's
+    // files and recreate all the `PathBuf`s.
+    pub fn search<F>(&self, cached_file_paths: &mut CachedFilePaths, mut pick: F)
         where F: FnMut(&Path, PathKind) -> FileMatch
     {
         self.for_each_lib_search_path(|lib_search_path, kind| {
             debug!("searching {}", lib_search_path.display());
-            let files = match fs::read_dir(lib_search_path) {
-                Ok(files) => files,
-                Err(..) => return,
-            };
-            let files = files.filter_map(|p| p.ok().map(|s| s.path()))
-                             .collect::<Vec<_>>();
+
+            // If we've previously gotten the paths within this directory,
+            // reuse our cached copy. Otherwise, get them and cache them.
+            let files = cached_file_paths.entry(lib_search_path.to_path_buf()).or_insert_with(|| {
```

Comment:
I think we should move the caching to `for_each_lib_search_path`, and compute all possible paths at once. I doubt we only need *some* of them, sometimes.

It's possible we could move the caching to constructing the `FileSearch`.